### PR TITLE
Add exponential fitting for ZNE

### PR DESCRIPTION
**Pull request by natestemen**
**Context:** As part of increasing the utility of ZNE within Catalyst, we'd like to add the option to use exponential fitting with ZNE. The issue 

**Description of the Change:** A function which is written using JAX is added which performs an exponential fit of data, and evaluates said fit at 0.

**Benefits:** Increased capabilities for users to play with. This is especially important given the benefits of tuning ZNE to the problem and backend at hand.

**Possible Drawbacks:** The current implementation transforms data using the natural logarithm $\ln$ to linearize the data so that a linear fit can be applied to learn the necessary parameters (see section below on the exponential model). Once they transformed parameters are learned, they can be converted back into those from the exponential model quite simply. The major drawback of this approach is that they expectation values have to be shifted to be in the domain of $\ln$ (i.e. $> 0$), or thrown away if small enough (see $\varepsilon$ `eps` parameter introduced in `expontial_extrapolate`).

**Related GitHub Issues:** https://github.com/PennyLaneAI/catalyst/issues/754

**Exponential model:** Assuming an exponential model with asymptote $y(x) = A\mathrm{e}^{Bx} + C$ we must first turn this into a linear equation to be able to apply a linear fit.

$$y(x) = A\mathrm{e}^{Bx} + C$$

$$y(x) - C = A\mathrm{e}^{Bx}$$

$$\ln(y(x) - C) = \ln(A) + Bx$$

$$\tilde{y}(x) = \tilde{A} + Bx$$

Once $\tilde{A}$ is learned, the hard part is over as the ZNE value is "simply" $y(0) = A + C$.

### CODE REVIEW

File: pennylane/transforms/mitigate.py
Comment (cosenal) [inline pennylane/transforms/mitigate.py:18]: It looks like jax in a optional dependency of pennylane, which means we should also handle the case when jax is not in the environment, e.g., https://github.com/PennyLaneAI/pennylane/blob/e63362d0a83ae412436cebcb6ca7d13e5c6f3579/pennylane/capture/primitives.py#L23-L27

**Comment by josh146** [comment]
Thanks for this PR @natestemen!

> ⚠️ draft PR note: I'm currently having trouble getting tests to pass on this change. I am aware and actively working on it. Putting this PR up for visibility and in case anyone has clues where I'm messing things up.

Feel free to tag us directly on the PR, or alternatively let us know if you want anyone to go through the PR.

File: pennylane/transforms/mitigate.py
Comment (natestemen) [reply pennylane/transforms/mitigate.py]: Good catch! Done in https://github.com/PennyLaneAI/pennylane/pull/5972/commits/88cde3b35ec09e0f2a3250438573be67adea849b.

**Comment by natestemen** [comment]
Guidance would be appreciated on resolving the CI error raised [here](https://github.com/PennyLaneAI/pennylane/actions/runs/9903181919/job/27358295670?pr=5972#step:8:1302).

**Comment by natestemen** [comment]
@josh146 ready for the first look!

File: pennylane/transforms/mitigate.py
Comment (rmoyard) [inline pennylane/transforms/mitigate.py:327]: In order to make it useful for different machine learning frameworks, you can replace jax.numpy with pennylane qml.math. Then users can use the function with Pytorch, TF, autograd.

File: pennylane/transforms/mitigate.py
Comment (rmoyard) [inline pennylane/transforms/mitigate.py:328]: Could you add some more docs about exponential extrapolate, your PR description would be useful here.

File: tests/transforms/test_mitigate.py
Comment (rmoyard) [inline tests/transforms/test_mitigate.py:135]: In this test file can you parametrize the extrapolation then we cover both?

File: tests/transforms/test_mitigate.py
Comment (rmoyard) [inline tests/transforms/test_mitigate.py:165]:
```suggestion
```

File: tests/transforms/test_mitigate.py
Comment (rmoyard) [inline tests/transforms/test_mitigate.py:526]: Could parametrize the test (different values for the numeric parameters)?

File: tests/transforms/test_mitigate.py
Comment (rmoyard) [inline tests/transforms/test_mitigate.py:150]:
```suggestion
```

**Review by rmoyard — COMMENTED** [review]
Thanks @natestemen, it looks great! I suggest you use the math module from PennyLane then the function is compatible with all frameworks and add some tests for the different ML frameworks. Otherwise make sure to parametize the unit test for the fit.

File: tests/transforms/test_mitigate.py
Comment (rmoyard) [inline tests/transforms/test_mitigate.py:557]: @natestemen When you add test that uses a specific framework you need to decorate with the right ML framework tag. The four tags are:
```
@pytest.mark.autograd
@pytest.mark.tf
@pytest.mark.torch
@pytest.mark.jax
```

File: pennylane/transforms/mitigate.py
Comment (cosenal) [inline pennylane/transforms/mitigate.py:30]: @rmoyard Do I understand it correctly from your review that once `qml.math` is used, this is no longer needed?

File: pennylane/transforms/mitigate.py
Comment (cosenal) [inline pennylane/transforms/mitigate.py:345]: why is https://github.com/PennyLaneAI/pennylane/pull/5972/files#diff-0b068d11a9b273faeddfe1ca83bdb5276086af16547f5fa308aa397fa184e135R321 a `tensor()` and this is an `Array()`?

File: tests/transforms/test_mitigate.py
Comment (cosenal) [inline tests/transforms/test_mitigate.py:526]: Technically this is testing extrapolation rather than fitting (i.e. ignores B)

File: pennylane/transforms/mitigate.py
Comment (natestemen) [reply pennylane/transforms/mitigate.py:327]: Cool! Out of curiosity where does `pennylane.math.sign` come from? I can't find it in the [`pennylane/math/__init__.py`](https://github.com/PennyLaneAI/pennylane/blob/48eef6c1c17dc69e5d467a46d5af337c85641fe0/pennylane/math/__init__.py) file.

I haven't figured out to how replace the line
```
y = jnp.array(y)
```
just yet. I tried `qml.math.array` and JAX complained with
```
E       jax.errors.TracerArrayConversionError: The numpy.ndarray conversion method __array__() was called on traced array with shape float64[].
E       The error occurred while tracing the function qfunc at /Users/nate/code/pennylane/tests/transforms/test_mitigate.py:459 for jit. This concrete value was not available in Python because it depends on the value of the argument theta.
E       See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerArrayConversionError
E       --------------------
E       For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.

/opt/homebrew/Caskroom/miniforge/base/envs/pennylane/lib/python3.11/site-packages/autoray/autoray.py:81: TracerArrayConversionError
```

I **did** get it to work with the following change, but it seems like an "incorrect" way to go about solving this problem.
```diff
-y = jnp.array(y)
+y = qml.math.stack(y)
```

Perhaps you have a recommendation?

File: tests/transforms/test_mitigate.py
Comment (natestemen) [reply tests/transforms/test_mitigate.py]: Good call. Done in https://github.com/PennyLaneAI/pennylane/pull/5972/commits/31aa0e4bd85d4948963e4311905aa9755734682d.

File: tests/transforms/test_mitigate.py
Comment (natestemen) [reply tests/transforms/test_mitigate.py]: Sorry about leaving these in! Removed here, and the other print calls in https://github.com/PennyLaneAI/pennylane/pull/5972/commits/58d1415a5ed0e2e3c5a38f0dd1a8818d532faa16.

File: tests/transforms/test_mitigate.py
Comment (natestemen) [reply tests/transforms/test_mitigate.py]: True. Renamed in https://github.com/PennyLaneAI/pennylane/pull/5972/commits/52825e1353853d5d67a6481608b0810a284ea083.

File: pennylane/transforms/mitigate.py
Comment (natestemen) [reply pennylane/transforms/mitigate.py:345]: I'm not totally sure, but I would think it has something to do with the fact that I'm using functionality outside of `poly_extrapolate`. Maybe once I convert everything to use `qml.math` this issue will go away.

All that said, when I run the example in `richardson_extrapolate`, I see a slightly different output than what we see in the docstring.
```python
>>> x = np.linspace(1, 10, 5)
>>> y = x**2 + x + 1 + 0.3 * np.random.rand(len(x))
>>> qml.transforms.richardson_extrapolate(x, y)
1.4828971589996394
```

File: tests/transforms/test_mitigate.py
Comment (natestemen) [reply tests/transforms/test_mitigate.py:553]: I hopefully shouldn't need this though, right? I don't think we want this functionality to only work for a subset of the frame works you listed, do we?

File: pennylane/transforms/mitigate.py
Comment (natestemen) [reply pennylane/transforms/mitigate.py]: Yup, meant to do that! Thanks: https://github.com/PennyLaneAI/pennylane/pull/5972/commits/113aaa881e39854c3032dceefcb8588c5927d3b3

File: pennylane/transforms/mitigate.py
Comment (josh146) [reply pennylane/transforms/mitigate.py:327]: > Cool! Out of curiosity where does `pennylane.math.sign` come from? I can't find it in the pennylane/math/__init__.py file.

@natestemen we're using [autoray](https://github.com/jcmgray/autoray) to automatically dispatch most of `qml.math.*` to the respective framework :) There are a few we needed to define ourselves either because autoray didn't support them, or there were bugs.

Comment (rmoyard) [reply pennylane/transforms/mitigate.py:327]: Is y not already a Jax array? When you jit the function, you cannot cast abstract type to Jax array. What happens if you do nothing on y?

Comment (natestemen) [reply pennylane/transforms/mitigate.py:327]: It is not. In `mitigate_with_zne` the results are stored in a python list:
https://github.com/PennyLaneAI/pennylane/blob/3bae4a2c2ea69e7734be8fbeb7641ae7cfe51a52/pennylane/transforms/mitigate.py#L537-L543

Thus, when I try to subtract a float (the asymptote) from the `y` values, I get the following error.
```
>       y_shifted = sign * (y - asymptote)
E       TypeError: unsupported operand type(s) for -: 'list' and 'float'
```

Comment (rmoyard) [reply pennylane/transforms/mitigate.py:327]: Then maybe what would work is to check if it is a list and if it is the case apply array otherwise not. Also you can probably keep the stack, if tests are not failing.

File: tests/transforms/test_mitigate.py
Comment (rmoyard) [reply tests/transforms/test_mitigate.py:547]: Generally we create a single test for each framework as it is easier to set up rather than have to cast parameters to each framework.

File: pennylane/transforms/mitigate.py
Comment (natestemen) [reply pennylane/transforms/mitigate.py:321]: I've used `pennylane.math.stack` since it allows us to get rid of the JAX dependency for exponential fitting (https://github.com/PennyLaneAI/pennylane/pull/5972/commits/b60ffc442cc71c1bc6152152edd4638d12dde9e3).

I tested to ensure it works properly on lists and it does, but it's hard to tell what it's doing on more complicated when before `stack` is called we have
```
[Traced<ConcreteArray(2.1213203435596397, dtype=float64)>with<JVPTrace(level=2/0)> with
  primal = Array(2.12132034, dtype=float64)
  tangent = Traced<ShapedArray(float64[])>with<JaxprTrace(level=1/0)> with
    pval = (ShapedArray(float64[]), None)
    recipe = JaxprEqnRecipe(eqn_id=<object object at 0x1749088b0>, in_tracers=(Traced<ShapedArray(float64[1]):JaxprTrace(level=1/0)>,), out_tracer_refs=[<weakref at 0x17e43c090; to 'JaxprTracer' at 0x17e43c040>], out_avals=[ShapedArray(float64[])], primitive=pjit, params={'jaxpr': { lambda ; a:f64[1]. let
    b:f64[] = reduce_sum[axes=(0,)] a
    c:f64[] = div b 1.0
  in (c,) }, 'in_shardings': (UnspecifiedValue,), 'out_shardings': (UnspecifiedValue,), 'in_layouts': (None,), 'out_layouts': (None,), 'resource_env': None, 'donated_invars': (False,), 'name': '_mean', 'keep_unused': False, 'inline': True}, effects=set(), source_info=SourceInfo(traceback=<jaxlib.xla_extension.Traceback object at 0x1698bce00>, name_stack=NameStack(stack=(Transform(name='jvp'),))), ctx=JaxprEqnContext(compute_type=None,threefry_partitionable=False)), Traced<ConcreteArray(2.120866995454891, dtype=float64)>with<JVPTrace(level=2/0)> with
  primal = Array(2.120867, dtype=float64)
  tangent = Traced<ShapedArray(float64[])>with<JaxprTrace(level=1/0)> with
    pval = (ShapedArray(float64[]), None)
    recipe = JaxprEqnRecipe(eqn_id=<object object at 0x174909000>, in_tracers=(Traced<ShapedArray(float64[1]):JaxprTrace(level=1/0)>,), out_tracer_refs=[<weakref at 0x17e43c3b0; to 'JaxprTracer' at 0x17e43c360>], out_avals=[ShapedArray(float64[])], primitive=pjit, params={'jaxpr': { lambda ; a:f64[1]. let
    b:f64[] = reduce_sum[axes=(0,)] a
    c:f64[] = div b 1.0
  in (c,) }, 'in_shardings': (UnspecifiedValue,), 'out_shardings': (UnspecifiedValue,), 'in_layouts': (None,), 'out_layouts': (None,), 'resource_env': None, 'donated_invars': (False,), 'name': '_mean', 'keep_unused': False, 'inline': True}, effects=set(), source_info=SourceInfo(traceback=<jaxlib.xla_extension.Traceback object at 0x1698c5970>, name_stack=NameStack(stack=(Transform(name='jvp'),))), ctx=JaxprEqnContext(compute_type=None,threefry_partitionable=False)), Traced<ConcreteArray(2.1206403214025205, dtype=float64)>with<JVPTrace(level=2/0)> with
  primal = Array(2.12064032, dtype=float64)
  tangent = Traced<ShapedArray(float64[])>with<JaxprTrace(level=1/0)> with
    pval = (ShapedArray(float64[]), None)
    recipe = JaxprEqnRecipe(eqn_id=<object object at 0x17e45ed80>, in_tracers=(Traced<ShapedArray(float64[1]):JaxprTrace(level=1/0)>,), out_tracer_refs=[<weakref at 0x17e43c6d0; to 'JaxprTracer' at 0x17e43c680>], out_avals=[ShapedArray(float64[])], primitive=pjit, params={'jaxpr': { lambda ; a:f64[1]. let
    b:f64[] = reduce_sum[axes=(0,)] a
    c:f64[] = div b 1.0
  in (c,) }, 'in_shardings': (UnspecifiedValue,), 'out_shardings': (UnspecifiedValue,), 'in_layouts': (None,), 'out_layouts': (None,), 'resource_env': None, 'donated_invars': (False,), 'name': '_mean', 'keep_unused': False, 'inline': True}, effects=set(), source_info=SourceInfo(traceback=<jaxlib.xla_extension.Traceback object at 0x16a5c79f0>, name_stack=NameStack(stack=(Transform(name='jvp'),))), ctx=JaxprEqnContext(compute_type=None,threefry_partitionable=False))]
```
and after we have
```
Traced<ConcreteArray([2.12132034 2.120867   2.12064032], dtype=float64)>with<JVPTrace(level=2/0)> with
  primal = Array([2.12132034, 2.120867  , 2.12064032], dtype=float64)
  tangent = Traced<ShapedArray(float64[3])>with<JaxprTrace(level=1/0)> with
    pval = (ShapedArray(float64[3]), None)
    recipe = JaxprEqnRecipe(eqn_id=<object object at 0x17e45eec0>, in_tracers=(Traced<ShapedArray(float64[1]):JaxprTrace(level=1/0)>, Traced<ShapedArray(float64[1]):JaxprTrace(level=1/0)>, Traced<ShapedArray(float64[1]):JaxprTrace(level=1/0)>), out_tracer_refs=[<weakref at 0x17e43d210; to 'JaxprTracer' at 0x17e43d1c0>], out_avals=[ShapedArray(float64[3])], primitive=concatenate, params={'dimension': 0}, effects=frozenset(), source_info=SourceInfo(traceback=<jaxlib.xla_extension.Traceback object at 0x1698c8cf0>, name_stack=NameStack(stack=(Transform(name='jvp'),))), ctx=JaxprEqnContext(compute_type=None,threefry_partitionable=False))
```

File: tests/transforms/test_mitigate.py
Comment (natestemen) [reply tests/transforms/test_mitigate.py:547]: Sorry, I need to take a step back. I don't think I'm adding any tests that are for a specific framework. At least that is the goal. Do I need to take any action here?

File: pennylane/transforms/mitigate.py
Comment (natestemen) [reply pennylane/transforms/mitigate.py]: Removed in https://github.com/PennyLaneAI/pennylane/pull/5972/commits/b60ffc442cc71c1bc6152152edd4638d12dde9e3.

File: tests/transforms/test_mitigate.py
Comment (rmoyard) [reply tests/transforms/test_mitigate.py:547]: Yes now that it works with all framework you can unit test it. For example create a  test `@pytest.mark.torch` `def test_extrapolate_exponential_torch` then inside this test you add a unit test that uses torch tensors. Repeat for the other frameworks. You can take a look in the test suite, you can check tests with those markers in the test suite. It should help a lot

File: pennylane/transforms/mitigate.py
Comment (rmoyard) [reply pennylane/transforms/mitigate.py:321]: That looks okay to me :+1:

File: tests/transforms/test_mitigate.py
Comment (natestemen) [reply tests/transforms/test_mitigate.py:585]: Added some further unit tests in https://github.com/PennyLaneAI/pennylane/pull/5972/commits/5248c02227f6672c64c6d420abb8c5108e29f107.

File: pennylane/transforms/mitigate.py
Comment (natestemen) [reply pennylane/transforms/mitigate.py]: Switching to `qml.math` means the output is slightly different now. Updated in https://github.com/PennyLaneAI/pennylane/pull/5972/commits/a7ec79755a98716c67672b71398c440d566e675b.

File: tests/transforms/test_mitigate.py
Comment (natestemen) [reply tests/transforms/test_mitigate.py]: Good idea. I added tests for both increasing and decreasing exponentials in https://github.com/PennyLaneAI/pennylane/pull/5972/commits/32af13c85c3c160844ba23013996c89b9c58baf8.

**Review by rmoyard — APPROVED** [review]
It looks good to me! The failure seems unrelated. We need two approvals, I will tag someone from the core team.

File: pennylane/transforms/mitigate.py
Comment (EmilianoG-byte) [inline pennylane/transforms/mitigate.py:329]: For clarity, should we mention that this is the `C` you introduced in the equation above?

File: pennylane/transforms/mitigate.py
Comment (EmilianoG-byte) [inline pennylane/transforms/mitigate.py:331]:
```suggestion
        eps (float): Epsilon to regularize :math:`\log(y - \text{asymptote})` when the argument is to close to
```

File: pennylane/transforms/mitigate.py
Comment (EmilianoG-byte) [inline pennylane/transforms/mitigate.py:337]:
```suggestion
    .. seealso:: :func:`~.pennylane.transforms.richardson_extrapolate`, :func:`~.pennylane.transforms.mitigate_with_zne`.
```

File: pennylane/transforms/mitigate.py
Comment (EmilianoG-byte) [inline pennylane/transforms/mitigate.py:344]:
```suggestion
    **Example:**

    >>> np.random.seed(0)
    >>> x = np.linspace(1, 10, 5)
    >>> y = np.exp(-x) + np.random.normal(scale=0.1, size=len(x))
    >>> qml.transforms.exponential_extrapolate(x, y)
    0.23365009000522544
```

I was thinking we can add a seed here for reproducibility.

**Review by EmilianoG-byte — COMMENTED** [review]
Great addition 🤩 ! Just left some comments regarding typos/cosmetics.

**Comment by EmilianoG-byte** [comment]
Is there a reason why we are not checking differentiability for other backends besides `jax`? For instance here:

https://github.com/PennyLaneAI/pennylane/blob/169d53e8c680572e8bc1dd420d0e5296c71174bb/tests/transforms/test_mitigate.py#L646-L647

Why didn't we add

```python
@pytest.mark.parametrize("extrapolate", [richardson_extrapolate, exponential_extrapolate])
```

like we did with JAX and JAX-JIT?

**Comment by natestemen** [comment]
Thanks for the docstring suggestions! Addressed in https://github.com/PennyLaneAI/pennylane/pull/5972/commits/e0815dac01fd235d8177b2f36d45e0c927cf284f.

> Is there a reason why we are not checking differentiability for other backends besides jax?

The goal of this PR was to get an exponential fit function in that would be compatible with [`pennylaneai/catalyst`](https://github.com/pennylaneai/catalyst). Originally I thought the `exponential_extrapolate` function was required to be written in JAX and hence only ensured it was tested as such. I wasn't aware of the testing requirements of the other backends[^1] until getting some review on this PR.

Happy to add your suggestion. Let me know.

[^1]: I'm still unclear if the other backends (torch, tf, autograd) are required in order for this functionality to work with catalyst. Is it?

**Comment by EmilianoG-byte** [comment]
> 1. I'm still unclear if the other backends (torch, tf, autograd) are required in order for this functionality to work with catalyst. Is it? [↩](#user-content-fnref-1-f78a22c9d9339f481c026149f1235a68)

Good question @natestemen 🤔 

Strictly speaking, the other interfaces are not needed for this functionality to work with Catalyst. However, since this will be part of the Pennylane features (even without Jitting our workflows), it should be compatible with the interfaces supported by Pennylane - this includes being able to differentiate it. 

The good thing is that since you are using `qml.math` in `exponential_extrapolate`, adding this support should be rather straightforward by parametrizing the differentiability tests with 

```python 
@pytest.mark.parametrize("extrapolate", [richardson_extrapolate, exponential_extrapolate])
```

wherever only `richardson_extrapolate` was used previously. Let me know if there is any unforeseen difficulty 😃.

File: pennylane/transforms/mitigate.py
Comment (EmilianoG-byte) [inline pennylane/transforms/mitigate.py:330]:
```suggestion
        asymptote (float): Infinite noise limit expected for your circuit of interest 
            (:math:`C` in the equation above). Defaults to 0 in the case an asymptote is not supplied.
```


This was rendering weirdly without the extra indentation

**Comment by rmoyard** [comment]
@natestemen The requirement for it to work with Jax is to be Jax-jittable. I have asked for other interfaces because in PL we always make things compatible for the different frameworks.

File: pennylane/transforms/mitigate.py
Comment (natestemen) [reply pennylane/transforms/mitigate.py]: good catch! done in https://github.com/PennyLaneAI/pennylane/pull/5972/commits/5e7867388ab0bfad6e07ff1437676f3a21e71462.

File: doc/releases/changelog-dev.md
Comment (rmoyard) [inline doc/releases/changelog-dev.md:28]: @natestemen Please add yourself in the list of contributors :)

**Review by EmilianoG-byte — APPROVED** [review]
Looks good to me! 😄

File: doc/releases/changelog-dev.md
Comment (dime10) [reply doc/releases/changelog-dev.md:33]: https://github.com/PennyLaneAI/pennylane/pull/5972/commits/e8c95e7cdbc71ae4bee4d1078b40410940d9ac49

**Pull request merged**
